EndpointSlice with Pods without an existing Node

aojea opened the pull request:
/kind bug
/kind cleanup
/kind documentation

#### What this PR does / why we need it:

endpointslices: node missing on Pod scenario

When a Pod is referencing a Node that doesn't exist on the local
informer cache, the current behavior was to return an error to
retry later and stop processing.
However, this can cause scenarios that a missing node leaves a
Slice stuck, it can no reflect other changes, or be created.
Also, this doesn't respect the publishNotReadyAddresses options
on Services, that considers ok to publish pod Addresses that are
known to not be ready.

The new behavior keeps retrying the problematic Service, but it
keeps processing the updates, reflacting current state on the
EndpointSlice. If the publishNotReadyAddresses is set, a missing
node on a Pod is not treated as an error.

```release-note
EndpointSlices with Pod referencing Nodes that doesn't exist couldn't be created or updated.
The behavior on the EndpointSlice controller has been modified to update the EndpointSlice without the Pods that reference non-existing Nodes, and keep retrying until all Pods reference existing Nodes.
However, if service.Spec.PublishNotReadyAddresses is set, all the Pods are published without retrying.
Fixed EndpointSlices metrics to reflect correctly the number of desired EndpointSlices when no endpoints are present.
```


Fixes: #107927

aojea committed:
endpointslices: node missing on Pod scenario

When a Pod is referencing a Node that doesn't exist on the local
informer cache, the current behavior was to return an error to
retry later and stop processing.
However, this can cause scenarios that a missing node leaves a
Slice stuck, it can no reflect other changes, or be created.
Also, this doesn't respect the publishNotReadyAddresses options
on Services, that considers ok to publish pod Addresses that are
known to not be ready.

The new behavior keeps retrying the problematic Service, but it
keeps processing the updates, reflacting current state on the
EndpointSlice. If the publishNotReadyAddresses is set, a missing
node on a Pod is not treated as an error.

aojea committed:
fix a bug on endpointslices tests comparing the wrong metrics

aojea commented:
/assign @thockin @robscott

dcbw commented on pkg/controller/endpointslice/metrics/cache.go:
```diff
@@ -89,7 +89,13 @@ func (spc *ServicePortCache) totals(maxEndpointsPerSlice int) (int, int, int) {
 		actualSlices += eInfo.Slices
-		desiredSlices += numDesiredSlices(eInfo.Endpoints, maxEndpointsPerSlice)
+		if eInfo.Endpoints > 0 {
+			desiredSlices += numDesiredSlices(eInfo.Endpoints, maxEndpointsPerSlice)
```
Also, do we need to change endpointslicemirroring at all?

aojea replied:
yeah, I think that is a matter of where we want to special-case the case "there is always one slice, despite there are not endpoints at all" ... I took this approach because `numDesiredSlices` is also used in

https://github.com/aojea/kubernetes/blob/ee08c977fd2e44ceb10506290fbaaf6b242d6a99/pkg/controller/endpointslicemirroring/metrics/cache.go#L139

aojea replied:
hmm, seems you are right

aojea replied:
th eendpointslicemirroring seems to not use a placeholder, and already returns 0 slices for 0 endpoints, but better @robscott for confirming

aojea committed:
fix metrics for placeholder slice

There is always a placeholder slice.

The ServicePortCache logic was considering always one endpointSlice
per Endpoint, but if there are multiple empty Endpoints, we just
use one placeholder slice, not multiple placeholder slices.

dcbw commented:
@aojea the conformance test failure looks odd, because it's about EPS. However, it doesn't look related to EndpointSlice at all; the failure is that the test pods don't get their IPs within a 3m timeout. TLDR the image pulls are taking 2+ minutes for the pod:

```
Jun 22 08:09:18 kind-worker kubelet[280]: I0622 08:09:18.943215     280 status_manager.go:685] "Patch status for pod" pod="endpointslice-805/pod1" patch="{\"metadata\":{\"uid\":\"61a94b34-adbb-4fc7-8429-8bec09
03d5ed\"},\"status\":{\"$setElementOrder/conditions\":[{\"type\":\"Initialized\"},{\"type\":\"Ready\"},{\"type\":\"ContainersReady\"},{\"type\":\"PodScheduled\"}],\"conditions\":[{\"lastProbeTime\":null,\"last
TransitionTime\":\"2022-06-22T08:09:11Z\",\"status\":\"True\",\"type\":\"Initialized\"},{\"lastProbeTime\":null,\"lastTransitionTime\":\"2022-06-22T08:09:11Z\",\"message\":\"containers with unready status: [co
ntainer1]\",\"reason\":\"ContainersNotReady\",\"status\":\"False\",\"type\":\"Ready\"},{\"lastProbeTime\":null,\"lastTransitionTime\":\"2022-06-22T08:09:11Z\",\"message\":\"containers with unready status: [con
tainer1]\",\"reason\":\"ContainersNotReady\",\"status\":\"False\",\"type\":\"ContainersReady\"}],\"containerStatuses\":[{\"image\":\"registry.k8s.io/e2e-test-images/nginx:1.14-2\",\"imageID\":\"\",\"lastState\
":{},\"name\":\"container1\",\"ready\":false,\"restartCount\":0,\"started\":false,\"state\":{\"waiting\":{\"reason\":\"ContainerCreating\"}}}],\"hostIP\":\"172.18.0.4\",\"startTime\":\"2022-06-22T08:09:11Z\"}}
"
```
```
Jun 22 08:12:05 kind-worker kubelet[280]: I0622 08:12:05.273805     280 event.go:294] "Event occurred" object="endpointslice-805/pod1" fieldPath="spec.containers{container1}" kind="Pod" apiVersion="v1" type="Normal" reason="Pulled" message="Successfully pulled image \"registry.k8s.io/e2e-test-images/nginx:1.14-2\" in 2m51.156327184s"
```

and a few seconds after that it times out.

dcbw commented:
/retest
